### PR TITLE
feat: Add support for EKS Auto Mode and EKS Hybrid nodes

### LINKED ISSUE #3226: please update EKS Automode.
## Is your request related to a new offering from AWS?
Is this functionality available in the [AWS provider for Terraform](https://registry.terraform.io/providers/hashicorp/aws/latest/docs)? See [CHANGELOG.md](https://github.com/hashicorp/terraform-provider-aws/blob/main/CHANGELOG.md), too.
  - No 🛑: please wait to file a request until the functionality is avaialble in the AWS provider
  - Yes ✅: please list the AWS provider version which introduced this functionality

## Is your request related to a problem? Please describe.

- https://aws.amazon.com/ko/blogs/aws/streamline-kubernetes-cluster-management-with-new-amazon-eks-auto-mode/?trk=d57158fd-77e3-423f-9e1e-005fd2a64d89&sc_channel=el
- https://docs.aws.amazon.com/ko_kr/eks/latest/userguide/automode.html

eks module doesn't support EKS AutoMode.

### PR DESCRIPTION
**Pull request by bryantbiggs**
## Description
- Add support for EKS Auto Mode
  - Adds new role creation for EKS Auto Mode node role when using built-in nodepools `system` and/or `general-purpose`
- Add support for EKS Hybrid nodes
- Add variable `enable_security_groups_for_pods` to maintain current cluster IAM role policy behavior; this can be disabled for those not using [security groups for pods](https://docs.aws.amazon.com/eks/latest/userguide/security-iam-awsmanpol.html#security-iam-awsmanpol-amazoneksvpcresourcecontroller). This variable and the attachment of the `AmazonEKSVPCResourceController` policy will be removed in the next breaking change

## Motivation and Context
- https://github.com/hashicorp/terraform-provider-aws/pull/40370
- https://github.com/hashicorp/terraform-provider-aws/pull/40371
- Resolves #3226

## Breaking Changes
- No

## How Has This Been Tested?
- [x] I have updated at least one of the `examples/*` to demonstrate and validate my change(s)
- [x] I have tested and validated these changes using one or more of the provided `examples/*` projects
- [x] I have executed `pre-commit run -a` on my pull request

### CODE REVIEW

**Comment by jatin-mehrotra-colorkrew** [comment]
Waiting to try this one.

File: README.md [hunk 1]
```diff
@@ -63,14 +202,30 @@ module "eks" {
...
+    Terraform   = "true"
+  }
+}
+```
+
+### Cluster Access Entry
+
+When enabling `authentication_mode = "API_AND_CONFIG_MAP"`, EKS will automatically create an access entry for the IAM role(s) used by managed node group(s) and Fargate profile(s). There are no additional actions required by users. For self-managed node groups and the Karpenter sub-module, this project automatically adds the access entry on behalf of users so there are no additional actions required by users.
+
+On clusters that were created prior to CAM support, there will be an existing access entry for the cluster creator. This was previously not visible when using `aws-auth` ConfigMap, but will become visible when access entry is enabled.
```
Comment (antonbabenko) [inline README.md:216]: What is CAM support?

File: examples/eks-hybrid-nodes/ami/amazon-eks-ubuntu.pkr.hcl [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+      "apt upgrade -y",
+      "apt install iptables conntrack -y",
+      "systemctl reboot",
+    ]
+
+    pause_after = "15s"
+  }
+
+  provisioner "shell" {
+    execute_command = "echo 'packer' | sudo -S sh -c '{{ .Vars }} {{ .Path }}'"
```
Comment (antonbabenko) [inline examples/eks-hybrid-nodes/ami/amazon-eks-ubuntu.pkr.hcl:289]: Should it really begin with - echo packer? Why not start it with sudo and not use pipe?

File: examples/eks-hybrid-nodes/main.tf [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+      protocol    = "tcp"
+      type        = "ingress"
+    }
+  }
+
+  node_security_group_additional_rules = {
+    hybrid-all = {
+      cidr_blocks = [local.remote_network_cidr]
+      description = "Allow all traffic from remote node/pod network"
+      from_port   = "-1"
```
Comment (antonbabenko) [inline examples/eks-hybrid-nodes/main.tf:60]: Can be a number like in the top block

**Review by antonbabenko — APPROVED** [review]
Looks very good and very detailed examples! 💪

My questions are minor and very much optional.

File: README.md (hunk 1 above)
Comment (bryantbiggs) [reply README.md:216]: CAM = cluster access management. I'll update to use the expanded form for clarity

File: examples/eks-hybrid-nodes/ami/amazon-eks-ubuntu.pkr.hcl (hunk 2 above)
Comment (bryantbiggs) [reply examples/eks-hybrid-nodes/ami/amazon-eks-ubuntu.pkr.hcl:289]: its Packer's way of executing commands as sudo when using a non-root role https://developer.hashicorp.com/packer/docs/provisioners/shell#sudo-example

File: README.md (hunk 1 above)
Comment (sergii-auctane) [reply README.md:216]: Cluster Auto Mode, i assume

**Comment by bryantbiggs** [comment]
FYI - https://github.com/hashicorp/terraform-provider-aws/issues/40411

File: examples/eks-hybrid-nodes/main.tf (hunk 3 above)
Comment (bryantbiggs) [reply examples/eks-hybrid-nodes/main.tf]: was able to remove these since the protocol is `all`

**Comment by bryantbiggs** [comment]
we'll go ahead and merge this since it supports creating clusters with EKS Auto Mode and EKS Hybrid nodes, and also supports opting into EKS Auto Mode - the only scenario which is not support at this time is opting into EKS Auto Mode utilizing the built-in nodepools (`system`, `general-purpose`). This requires https://github.com/hashicorp/terraform-provider-aws/issues/40411 to be resolved to support that scenario

File: README.md (hunk 1 above)
Comment (bryantbiggs) [reply README.md]: corrected in https://github.com/terraform-aws-modules/terraform-aws-eks/pull/3225/commits/c24470be1e0a8204beb7f6a6d15cd039ae44f5d8

**Comment by antonbabenko** [comment]
This PR is included in version 20.31.0 :tada:

**Pull request merged**